### PR TITLE
Added basic suite test

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary 

- This PR is an enhancement.
- It added a basic test to run in CI that runs all kernels and variants and verifies that they actually ran and that checksums are within 10e-7 tolerance of reference variant. We should be able to tighten this tolerance with a more robust checksum computation (see https://github.com/LLNL/RAJAPerf/pull/238). However, the test runs abbreviated versions of the kernels. The checksum diffs tend to get closer to zero as we run closer to the default.
- CI is enabled for azure and gitlab, although it could use some polishing.

### CODE REVIEW

**Comment by kab163** [comment]
> @kab163 i'll get to that after I change a few other things. Unfortunately, all recent xl versions have a bug in their openmp runtime, I believe, which causes all openmp variants of one kernel to generate garbage results

Uh Oh.. does that mean we need to exclude a specific subset of tests from running in the case that it's a xl+openmp job? (I'm thinking of how we did that for certain known-hip-failures for RAJA)

**Comment by rhornung67** [comment]
Probably.  I'll work on that after I get everything else working. Also, we found a bug in the RAJA CMake stuff that needs to be addressed before this PR can be merged. Working on that too.

File: test/test-raja-perf-suite.cpp
Comment (rhornung67) [inline test/test-raja-perf-suite.cpp:24]: @MrBurmark here's what I came up with to disable the failing fused kernel test for hip builds. It's gross and should be checked with new versions. Any thoughts for improvement?

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:24]: I think its reasonable for now, just have to add an issue to make it better later.
Would it be a bit simpler if you used sargv.size() instead of a separate argc?

File: test/test-raja-perf-suite.cpp
Comment (MrBurmark) [inline test/test-raja-perf-suite.cpp:39]: Would it make more sense to make executor take const char** instead? We'd have to add a cast in main, but it could avoid const_cast here.

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:39]: Or have two constructors for Executor to hide the cast of char** to const char** there.

File: test/test-raja-perf-suite.cpp
Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:24]: Can you use a newer rocm version, kernel fusion appears to work here with rocm 5.1.

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:24]: I used the argc/argv stuff from the start since it mimics the args to main, MPI_Init, etc. We would have to change the RunParams method that parses the command line arg list if we changed what we pass to the Executor ctor.

Of course, we can update the compiler versions. Let's do in a separate PR. Actually, that would be a good test to see if the RAJA dev guide CI docs I am working on are useful -- have someone see if they can do it following what's there.

File: test/test-raja-perf-suite.cpp
Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:39]: I'm not sure it makes much difference. Again, following the common practice....

`int main( int argc, char** argv )`

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:39]: Ya, const_cast is just evil.

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:39]: I was trying to make the code simpler using C++ strings. c_str() returns const char*

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:39]: Its fine, I was just hoping for an easy way to avoid the const_cast.

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:39]: i kinda like it. it makes me feel naughty.

File: test/test-raja-perf-suite.cpp
Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:24]: @MrBurmark rocm 5.1 is available on rznevada. We're not quite there yet with CI testing for RAJA/RAJAPerf. @kab163 is working on it. On corona, the latest available is rocm 5.0.2.

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:24]: We can use the HIP_VERSION to enable/disable the test so it automatically turns on when we can test it https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L261.

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:24]: So change the guard to this?

`#if (HIP_VERSION_MAJOR >= 5 && HIP_VERSION_MINOR >= 1)`

Comment (MrBurmark) [reply test/test-raja-perf-suite.cpp:24]: I think you want something more like this, or the negative of this actually in this case.

```
#if (HIP_VERSION_MAJOR > 5 || (HIP_VERSION_MAJOR == 5 && HIP_VERSION_MINOR >= 1))
```

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:24]: I thought the test fails for rocm 5.0.2?

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp:24]: Gotcha.

**Comment by rhornung67** [comment]
@MrBurmark This should be done now if you want to take a look.  Need to wait for PRs to be merged in RAJA (https://github.com/LLNL/RAJA/pull/1253) and radiuss-spack-configs (https://github.com/LLNL/radiuss-spack-configs/pull/31) and pull those in before we can merge this.

File: test/test-raja-perf-suite.cpp
Comment (MrBurmark) [inline test/test-raja-perf-suite.cpp:29]: I suppose these wouldn't overlap but should they still be separated to better match up with what is below?

Comment (rhornung67) [reply test/test-raja-perf-suite.cpp]: Change pushed.

**Comment by MrBurmark** [comment]
Looks pretty good, I'll take a final look after those PRs are in.

**Comment by rhornung67** [comment]
Folks, all PRs this depended on have been merged and submodules have been updated in this branch. So this should be good to merge when checks pass.  Please review. Thank you.

File: .gitlab/lassen-jobs.yml
Comment (MrBurmark) [inline .gitlab/lassen-jobs.yml:34]: Should we document why we're using -qstrict?

File: CMakeLists.txt
Comment (MrBurmark) [inline CMakeLists.txt:38]: Why is this done like this?

Comment (rhornung67) [reply CMakeLists.txt:38]: I wanted to turn off camp and RAJA tests because they take too long to build. I added a new variable to emphasize the point. It could be done other ways, I suppose. Is it confusing?

File: .gitlab/lassen-jobs.yml
Comment (rhornung67) [reply .gitlab/lassen-jobs.yml]: I thought it would be batter to describe in an issue, so we can try to see if it works at some point.  Although, Mike C. this sort of thing has been an issue with XL for many years.

File: CMakeLists.txt
Comment (MrBurmark) [reply CMakeLists.txt:38]: I mean why not use the commented out cmake_dependent_option?

Comment (rhornung67) [reply CMakeLists.txt:38]: Ah. I think I tried that and it didn't work for me. It may have been due to other breakage I was dealing with.

Comment (davidbeckingsale) [reply CMakeLists.txt:38]: It would be good to use the dependent option if possible.

File: CMakeLists.txt
Comment (davidbeckingsale) [inline CMakeLists.txt:44]: Do we have a `RAJA_ENABLE_EXAMPLES` option that can be used instead?

Comment (davidbeckingsale) [reply CMakeLists.txt:44]: Actually, looks like that is set below and this line can be deleted.

File: CMakeLists.txt
Comment (rhornung67) [reply CMakeLists.txt:38]: If I do that, I would have to set both ENABLE_TESTS and RAJA_PERFSUITE_ENABLE_TESTS to On? Or, am I mis-understanding?

File: CMakeLists.txt
Comment (rhornung67) [reply CMakeLists.txt:44]: will do. thanks.

File: test/CMakeLists.txt
Comment (davidbeckingsale) [inline test/CMakeLists.txt:9]: I think this should be a `target_include_directories` instead, e.g.:
```
target_include_directories(test-raja-perf-suite.exe PRIVATE ../src)
```

You could also use `PROJECT_SOURCE_DIR/src` to avoid using the relative path.

**Review by davidbeckingsale — DISMISSED** [review]
This is fine just a couple of CMake comments

File: CMakeLists.txt
Comment (davidbeckingsale) [reply CMakeLists.txt:38]: The dependent option would set `RAJA_PERFSUITE_ENABLE_TESTS` to `On` if `ENABLE_TESTS` is On - so you only have to set `ENABLE_TESTS`.

Comment (rhornung67) [reply CMakeLists.txt:38]: Thanks. Will try.

File: test/CMakeLists.txt
Comment (rhornung67) [reply test/CMakeLists.txt:9]: Using `PROJECT_SOURCE_DIR/src` the test doesn't build since it can't find the needed header files.  

`target_include_directories(test-raja-perf-suite.exe PRIVATE ../src)`  works if I put it after the `raja_add_test` macro call

Comment (rhornung67) [reply test/CMakeLists.txt:9]: Besides being more verbose, what is the advantage of this over what I had originally?

Comment (davidbeckingsale) [reply test/CMakeLists.txt:9]: `${PROJECT_SOURCE_DIR}/src` should work. It's mostly a best practice thing.

Comment (rhornung67) [reply test/CMakeLists.txt]: change pushed to use target_include_directories

**Comment by rhornung67** [comment]
@MrBurmark @artv3 @davidbeckingsale @kab163 I've updated XL and fixed the build flags and updated to ROCm 5.1.0 for Corona. All checks should pass now, I've turned off allowing failures for XL+CUDA. Please review  and approve if you think this is good to merge.

**Review by MrBurmark — APPROVED** [review]

**Pull request merged**
